pkg/storage/utils/decomposedfs/decomposedfs.go

```diff
@@ -552,20 +553,175 @@ func (fs *Decomposedfs) Download(ctx context.Context, ref *provider.Reference) (
+	}
+	defer f.Close()
+
+	if err := json.NewEncoder(f).Encode(lock); err != nil {
+		return errors.Wrap(err, "Decomposedfs: could not write lock file")
+	}
+
+	return nil
 }
 
 // RefreshLock refreshes an existing lock on the given reference
 func (fs *Decomposedfs) RefreshLock(ctx context.Context, ref *provider.Reference, lock *provider.Lock) error {
-	return errtypes.NotSupported("unimplemented")
+	node, err := fs.lu.NodeFromResource(ctx, ref)
+	if err != nil {
+		return errors.Wrap(err, "Decomposedfs: error resolving ref")
+	}
+
+	if !node.Exists {
+		return errtypes.NotFound(filepath.Join(node.ParentID, node.Name))
+	}
+
+	ok, err := fs.p.HasPermission(ctx, node, func(rp *provider.ResourcePermissions) bool {
+		return rp.InitiateFileUpload
+	})
+	switch {
+	case err != nil:
+		return errtypes.InternalError(err.Error())
+	case !ok:
+		return errtypes.PermissionDenied(filepath.Join(node.ParentID, node.Name))
+	}
+
+	lockPath := node.InternalPath() + ".lock"
+
+	f, err := os.OpenFile(lockPath, os.O_RDWR, os.ModeExclusive)
+	if err != nil {
+		return errors.Wrap(err, "Decomposedfs: could not open lock file")
+	}
+	defer f.Close()
+
+	oldLock := &provider.Lock{}
+	if err := json.NewDecoder(f).Decode(oldLock); err != nil {
+		return errors.Wrap(err, "Decomposedfs: could not read lock")
+	}
+
+	u := ctxpkg.ContextMustGetUser(ctx)
+	if !utils.UserEqual(oldLock.GetUser(), u.Id) {
+		return errtypes.PermissionDenied("cannot refresh lock of another holder")
+	}
+
+	if !utils.UserEqual(oldLock.GetUser(), lock.GetUser()) {
+		return errtypes.PermissionDenied("cannot change holder when refreshing a lock")
+	}
+
+	if err := json.NewEncoder(f).Encode(lock); err != nil {
+		return errors.Wrap(err, "Decomposedfs: could not write lock file")
+	}
+
+	return nil
 }
 
 // Unlock removes an existing lock from the given reference
 func (fs *Decomposedfs) Unlock(ctx context.Context, ref *provider.Reference) error {
-	return errtypes.NotSupported("unimplemented")
+	node, err := fs.lu.NodeFromResource(ctx, ref)
+	if err != nil {
+		return errors.Wrap(err, "Decomposedfs: error resolving ref")
+	}
+
+	if !node.Exists {
+		return errtypes.NotFound(filepath.Join(node.ParentID, node.Name))
+	}
+
+	ok, err := fs.p.HasPermission(ctx, node, func(rp *provider.ResourcePermissions) bool {
+		return rp.InitiateFileUpload // TODO do we need a dedicated permission?
+	})
+	switch {
+	case err != nil:
+		return errtypes.InternalError(err.Error())
+	case !ok:
+		return errtypes.PermissionDenied(filepath.Join(node.ParentID, node.Name))
+	}
+
+	lockPath := node.InternalPath() + ".lock"
+
+	f, err := os.OpenFile(lockPath, os.O_RDONLY, os.ModeExclusive)
+	if err != nil {
+		return errors.Wrap(err, "Decomposedfs: could not open lock file")
+	}
+
+	oldLock := &provider.Lock{}
+	if err := json.NewDecoder(f).Decode(oldLock); err != nil {
+		_ = f.Close()
+		return errors.Wrap(err, "Decomposedfs: could not read lock")
+	}
+
+	// TODO is the app in the context?
```

Comment:
@wkloucek @glpatcern where do I get the app from? or should I only check it when refreshing locks? Can the user be nil? Can there be only an app?

Reply:
Hey @butonic, good point: the case of user == nil and app != nil is legitimate and was thought as a way to represent the workflow with WOPI apps. But some `appname` context would be needed in the calls to SetLock / RefreshLock / Unlock to be checked against: I don't think we can "store" it in the `ctx`, rather it has to be passed to the functions, which would require a change of the CS3API protocol. @wkloucek what do you think?